File: apps/cli/src/main/kotlin/MediaValidation.kt
package com.sloopworks.dayfold.cli

// ADR 0036 — shared hardened validation for author-supplied image enrichment.
//
// MUST stay byte-for-byte in lock-step with the server (apps/api/src/media-
// validation.ts) and the client (apps/client/.../client/MediaValidation.kt) — a
// parser differential IS the vulnerability. Pure Kotlin, no platform deps, so the
// SAME source compiles in the client's commonMain. (Phase 2: codegen all three
// from one source.)
//
// Phase-1 posture: https-only, no userinfo, exact-host allowlist (NOT suffix),
// port empty/443, reject SVG, length-cap, curated-icon enum, #RRGGBB accent.
// Allowlist = exactly `upload.wikimedia.org`.
object MediaValidation {
  val ALLOWED_IMAGE_HOSTS: Set<String> = setOf("upload.wikimedia.org")

  // 18 curated icon NAMES from the signed-off design (Enrichment.dc.html §D).
  val CURATED_ICONS: Set<String> = setOf(
    "school", "luggage", "medical", "move", "party", "baby",
    "calendar", "location", "link", "document", "contact", "budget",
    "travel", "car", "food", "pet", "sport", "list",
  )

  private const val MAX_URL_LEN = 2048
  private val ACCENT_RE = Regex("^#[0-9a-fA-F]{6}$")

  /** null = acceptable, else a short reason. Hand-rolled to match the WHATWG-URL
   *  decisions the server reaches for the ASCII-only allowlist. */
  fun imageUrlError(url: String?): String? {
    if (url == null) return "must be a string"
    if (url.isEmpty() || url.length > MAX_URL_LEN) return "url length out of range"
    // reject control chars / whitespace / backslash before parsing (no smuggling).
    if (url.any { it.code <= 0x20 || it.code == 0x7f || it == '\\' }) return "url contains illegal characters"

    val schemeSep = url.indexOf("://")
    if (schemeSep < 0) return "url does not parse"
    if (url.substring(0, schemeSep).lowercase() != "https") return "scheme must be https" // blocks http/data/javascript/blob

    val rest = url.substring(schemeSep + 3)
    val authEnd = rest.indexOfFirst { it == '/' || it == '?' || it == '#' }
    val authority = if (authEnd < 0) rest else rest.substring(0, authEnd)
    val path = if (authEnd < 0) "" else rest.substring(authEnd)

    if (authority.contains('@')) return "userinfo not allowed"
    if (authority.contains('[') || authority.contains(']')) return "ip-literal host not allowed"

    var host = authority
    var port = ""
    val colon = authority.lastIndexOf(':')
    if (colon >= 0) { host = authority.substring(0, colon); port = authority.substring(colon + 1) }
    if (port.isNotEmpty() && port != "443") return "explicit port not allowed"

    host = host.lowercase().removeSuffix(".")
    // Exact-host match. Any non-ascii (IDN homograph) or stray char → illegal host,
    // so a unicode/punycode-decoded host can never alias the allowlisted ASCII host.
    if (host.isEmpty() || host.any { !(it in 'a'..'z' || it in '0'..'9' || it == '.' || it == '-') })
      return "illegal host"
    if (host !in ALLOWED_IMAGE_HOSTS) return "host \"$host\" is not on the image allowlist"

    val pathOnly = path.substringBefore('?').substringBefore('#').lowercase()
    if (pathOnly.endsWith(".svg")) return "SVG images are not allowed"
    return null
  }

  fun iconError(icon: String?): String? =
    if (icon == null) "must be a string"
    else if (icon in CURATED_ICONS) null else "icon \"$icon\" is not in the curated set"

  fun accentHexError(hex: String?): String? =
    if (hex == null) "must be a string"
    else if (ACCENT_RE.matches(hex)) null else "accentColor must match #RRGGBB"
}

File: apps/api/src/media-validation.ts
// ADR 0036 — shared hardened validation for author-supplied image enrichment.
//
// One rule, mirrored across server (this file), client
// (apps/client/.../client/MediaValidation.kt) and CLI
// (apps/cli/.../MediaValidation.kt). A parser differential IS the vulnerability,
// so keep the three in lock-step (Phase 2 will codegen them from one source).
//
// Phase-1 posture (ADR 0036): https-only, no userinfo, exact-host allowlist
// (NOT suffix), port empty/443, reject SVG, length-cap, curated-icon enum,
// #RRGGBB accent. The allowlist is exactly `upload.wikimedia.org`.

export const ALLOWED_IMAGE_HOSTS: ReadonlySet<string> = new Set(["upload.wikimedia.org"]);

// Curated icon NAMES (18) from the signed-off design (Enrichment.dc.html §D).
// Author sets the NAME; the client maps name→glyph; unknown → fallback tile.
export const CURATED_ICONS: ReadonlySet<string> = new Set([
  "school", "luggage", "medical", "move", "party", "baby",
  "calendar", "location", "link", "document", "contact", "budget",
  "travel", "car", "food", "pet", "sport", "list",
]);

const MAX_URL_LEN = 2048;
const ACCENT_RE = /^#[0-9a-fA-F]{6}$/;

export type MediaIssue = { path: (string | number)[]; message: string };

/**
 * Validate one author-supplied image URL against the ADR 0036 hardened rule.
 * Returns null when acceptable, else a short reason (caller wraps it in a 422
 * issue). Uses the WHATWG URL parser (IDNA/punycode + normalization) then layers
 * the explicit allowlist checks. The Kotlin mirrors apply identical logical steps.
 */
export function imageUrlError(url: unknown): string | null {
  if (typeof url !== "string") return "must be a string";
  if (url.length === 0 || url.length > MAX_URL_LEN) return "url length out of range";
  // reject control chars / whitespace / backslash before parsing (no smuggling).
  if (/[\x00-\x20\x7f\\]/.test(url)) return "url contains illegal characters";

  let u: URL;
  try { u = new URL(url); } catch { return "url does not parse"; }

  if (u.protocol !== "https:") return "scheme must be https";          // blocks http/data/javascript/blob
  if (u.username !== "" || u.password !== "") return "userinfo not allowed";
  if (u.port !== "") return "explicit port not allowed";               // 443 is normalized to ""
  // WHATWG already lowercased + punycode(toASCII)-normalized the host; strip a
  // single trailing dot, then require an EXACT host match (never endsWith).
  const host = u.hostname.replace(/\.$/, "");
  if (!ALLOWED_IMAGE_HOSTS.has(host)) return `host "${host}" is not on the image allowlist`;
  if (u.pathname.toLowerCase().endsWith(".svg")) return "SVG images are not allowed";
  return null;
}

export function iconError(icon: unknown): string | null {
  if (typeof icon !== "string") return "must be a string";
  return CURATED_ICONS.has(icon) ? null : `icon "${icon}" is not in the curated set`;
}

export function accentHexError(hex: unknown): string | null {
  if (typeof hex !== "string") return "must be a string";
  return ACCENT_RE.test(hex) ? null : "accentColor must match #RRGGBB";
}

// ---- object-level validators (return zod-issue-shaped lists) ----------------

function urlField(media: any, key: string, base: (string | number)[], out: MediaIssue[]) {
  if (media[key] == null) return;
  const e = imageUrlError(media[key]);
  if (e) out.push({ path: [...base, key], message: e });
}

/** Hub.media (heroUrl, thumbnailUrl, icon, accentColor). */
export function validateHubMedia(media: unknown): MediaIssue[] {
  if (media == null) return [];
  if (typeof media !== "object") return [{ path: ["media"], message: "must be an object" }];
  const m = media as any, out: MediaIssue[] = [];
  urlField(m, "heroUrl", ["media"], out);
  urlField(m, "thumbnailUrl", ["media"], out);
  if (m.icon != null) { const e = iconError(m.icon); if (e) out.push({ path: ["media", "icon"], message: e }); }
  if (m.accentColor != null) { const e = accentHexError(m.accentColor); if (e) out.push({ path: ["media", "accentColor"], message: e }); }
  return out;
}

/** BriefingCard.media (icon, accentColor, thumbnailUrl). */
export function validateCardMedia(media: unknown): MediaIssue[] {
  if (media == null) return [];
  if (typeof media !== "object") return [{ path: ["media"], message: "must be an object" }];
  const m = media as any, out: MediaIssue[] = [];
  urlField(m, "thumbnailUrl", ["media"], out);
  if (m.icon != null) { const e = iconError(m.icon); if (e) out.push({ path: ["media", "icon"], message: e }); }
  if (m.accentColor != null) { const e = accentHexError(m.accentColor); if (e) out.push({ path: ["media", "accentColor"], message: e }); }
  return out;
}

/** Block payload image fields: link/document thumbnailUrl, contact avatarUrl + accentColor. */
export function validateBlockPayloadMedia(type: unknown, payload: unknown): MediaIssue[] {
  if (payload == null || typeof payload !== "object") return [];
  const p = payload as any, out: MediaIssue[] = [];
  if (type === "link" || type === "document") urlField(p, "thumbnailUrl", ["payload"], out);
  if (type === "contact") {
    urlField(p, "avatarUrl", ["payload"], out);
    if (p.accentColor != null) { const e = accentHexError(p.accentColor); if (e) out.push({ path: ["payload", "accentColor"], message: e }); }
  }
  return out;
}

/** Lowercase the accent on write (in place on a shallow copy is the caller's job). */
export function normalizedAccent(hex: string | null | undefined): string | null {
  return typeof hex === "string" ? hex.toLowerCase() : null;
}
